### ISSUE TITLE
Installation issue: Nettle

### ISSUE THREAD

**Issue by ocfmatt** [spack/spack#11773]
### Steps to reproduce the issue

```console
$ spack install nettle %gcc@4.8.5
...
==> Error: ProcessError: Command exited with status 2:
    'make' '-j80'

3 errors found in build log:
     643    && true
     644    /opt/software/uoa/2019/spack/lib/spack/env/gcc/gcc -I.  -DHAVE_CONFIG_H -g -O2 -ggdb3 -Wno-pointer-sign -Wall -W   -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes   -Wpointer-arith -Wbad-function-cast -Wne
            sted-externs -fpic -MT dsa-sha256-sign.o -MD -MP -MF dsa-sha256-sign.o.d -c dsa-sha256-sign.c \
     645    && true
     646    rsa-sign-tr.c: In function 'sec_equal':
     647    /opt/software/uoa/2019/spack/lib/spack/env/gcc/gcc -I.  -DHAVE_CONFIG_H -g -O2 -ggdb3 -Wno-pointer-sign -Wall -W   -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes   -Wpointer-arith -Wbad-function-cast -Wne
            sted-externs -fpic -MT dsa-sha256-verify.o -MD -MP -MF dsa-sha256-verify.o.d -c dsa-sha256-verify.c \
     648    && true
  >> 649    rsa-sign-tr.c:243:3: error: 'for' loop initial declarations are only allowed in C99 mode
     650       for (size_t i = 0; i < limbs; i++)
     651       ^
     652    rsa-sign-tr.c:243:3: note: use option -std=c99 or -std=gnu99 to compile your code
     653    /opt/software/uoa/2019/spack/lib/spack/env/gcc/gcc -I.  -DHAVE_CONFIG_H -g -O2 -ggdb3 -Wno-pointer-sign -Wall -W   -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes   -Wpointer-arith -Wbad-function-cast -Wne
            sted-externs -fpic -MT dsa2sexp.o -MD -MP -MF dsa2sexp.o.d -c dsa2sexp.c \
     654    && true
     655    /opt/software/uoa/2019/spack/lib/spack/env/gcc/gcc -I.  -DHAVE_CONFIG_H -g -O2 -ggdb3 -Wno-pointer-sign -Wall -W   -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes   -Wpointer-arith -Wbad-function-cast -Wne
            sted-externs -fpic -MT sexp2dsa.o -MD -MP -MF sexp2dsa.o.d -c sexp2dsa.c \
     656    && true
  >> 657    make[1]: *** [rsa-sign-tr.o] Error 1
     658    make[1]: *** Waiting for unfinished jobs....
     659    make[1]: Leaving directory `/tmp/root/spack-stage/spack-stage-sHHgXL/nettle-3.4.1'
  >> 660    make: *** [all] Error 2

```

### Platform and user environment

Please report your OS here:
```commandline
$ uname -a 
Linux maxlogin2 3.10.0-957.10.1.el7.x86_64 #1 SMP Mon Mar 18 15:06:45 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
``` 

### Additional information

My target application is CP2K using Nettle as a dependency.

```console
$ ./spack spec --install-status nettle %gcc@4.8.5
Input spec
--------------------------------
 -   nettle%gcc@4.8.5

Concretized
--------------------------------
 -   nettle@3.4.1%gcc@4.8.5 arch=linux-centos7-x86_64
[+]      ^gmp@6.1.2%gcc@4.8.5 arch=linux-centos7-x86_64
[+]          ^autoconf@2.69%gcc@4.8.5 arch=linux-centos7-x86_64
[+]              ^m4@1.4.18%gcc@4.8.5 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,c0a408fbffb7255fcc75e26bd8edab116fc81d216bfd18b473668b7739a4158e,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 +sigsegv arch=linux-centos7-x86_64
[+]                  ^libsigsegv@2.11%gcc@4.8.5 arch=linux-centos7-x86_64
[+]              ^perl@5.26.2%gcc@4.8.5+cpanm patches=0eac10ed90aeb0459ad8851f88081d439a4e41978e586ec743069e8b059370ac +shared+threads arch=linux-centos7-x86_64
[+]                  ^gdbm@1.18.1%gcc@4.8.5 arch=linux-centos7-x86_64
[+]                      ^readline@7.0%gcc@4.8.5 arch=linux-centos7-x86_64
[+]                          ^ncurses@6.1%gcc@4.8.5~symlinks~termlib arch=linux-centos7-x86_64
[+]                              ^pkgconf@1.6.0%gcc@4.8.5 arch=linux-centos7-x86_64
[+]          ^automake@1.16.1%gcc@4.8.5 arch=linux-centos7-x86_64
[+]          ^libtool@2.4.6%gcc@4.8.5 arch=linux-centos7-x86_64

```

You might want to run this command on the `config.log` or any other similar file
found in the stage directory: 
```console
./spack location -s nettle
/opt/software/uoa/2019/spack/var/spack/stage/nettle-3.4.1-dzgj2ckkldnwi4e23k6s6zmbanbucbm3
```

```console
$ ./spack -d install -j 1 nettle
...
==> [2019-06-18-09:43:00.288740] Error: ProcessError: Command exited with status 2:
    'make'

3 errors found in build log:
     565    && true
     566    /opt/software/uoa/2019/spack/lib/spack/env/gcc/gcc -I.  -DHAVE_CONFIG_H -g -O2 -ggdb3 -Wno-pointer-sign -Wall -W   -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes   -Wpointer-arith -Wbad-function-cast -Wne
            sted-externs -fpic -MT rsa-sign.o -MD -MP -MF rsa-sign.o.d -c rsa-sign.c \
     567    && true
     568    /opt/software/uoa/2019/spack/lib/spack/env/gcc/gcc -I.  -DHAVE_CONFIG_H -g -O2 -ggdb3 -Wno-pointer-sign -Wall -W   -Wmissing-prototypes -Wmissing-declarations -Wstrict-prototypes   -Wpointer-arith -Wbad-function-cast -Wne
            sted-externs -fpic -MT rsa-sign-tr.o -MD -MP -MF rsa-sign-tr.o.d -c rsa-sign-tr.c \
     569    && true
     570    rsa-sign-tr.c: In function 'sec_equal':
  >> 571    rsa-sign-tr.c:243:3: error: 'for' loop initial declarations are only allowed in C99 mode
     572       for (size_t i = 0; i < limbs; i++)
     573       ^
     574    rsa-sign-tr.c:243:3: note: use option -std=c99 or -std=gnu99 to compile your code
  >> 575    make[1]: *** [rsa-sign-tr.o] Error 1
     576    make[1]: Leaving directory `/tmp/root/spack-stage/spack-stage-4Jz1N8/nettle-3.4.1'
  >> 577    make: *** [all] Error 2

See build log for details:
  /opt/software/uoa/2019/spack/var/spack/stage/nettle-3.4.1-dzgj2ckkldnwi4e23k6s6zmbanbucbm3/spack-build.out
Traceback (most recent call last):
  File "/opt/software/uoa/2019/spack/lib/spack/spack/build_environment.py", line 765, in child_process
    return_value = function()
  File "/opt/software/uoa/2019/spack/lib/spack/spack/package.py", line 1567, in build_process
    phase(self.spec, self.prefix)
  File "/opt/software/uoa/2019/spack/lib/spack/spack/package.py", line 99, in phase_wrapper
    phase(spec, prefix)
  File "/opt/software/uoa/2019/spack/lib/spack/spack/build_systems/autotools.py", line 263, in build
    inspect.getmodule(self).make(*self.build_targets)
  File "/opt/software/uoa/2019/spack/lib/spack/spack/build_environment.py", line 131, in __call__
    return super(MakeExecutable, self).__call__(*args, **kwargs)
  File "/opt/software/uoa/2019/spack/lib/spack/spack/util/executable.py", line 189, in __call__
    proc.returncode, long_msg)
ProcessError: Command exited with status 2:
    'make'
```

**Comment by ocfmatt** [comment]
[spack-cc-nettle-dzgj2ck.in.log](https://github.com/spack/spack/files/3300626/spack-cc-nettle-dzgj2ck.in.log)
[spack-cc-nettle-dzgj2ck.out.log](https://github.com/spack/spack/files/3300627/spack-cc-nettle-dzgj2ck.out.log)

**Comment by adamjstewart** [comment]
According to the error message:
```
error: 'for' loop initial declarations are only allowed in C99 mode
```
it sounds like the package needs to be built in C99 mode. Perhaps we need to add `self.compiler.c99_flag` to a certain flag?

**Comment by ocfmatt** [comment]
After more searching around and some sleep I've found there is an upstream fix in Nettle to remove the accidental use of a C99 loop.

See commit: https://github.com/gnutls/nettle/commit/f5a3a224bf00bef5669366d2ae23c2b2b13b8016

**Comment by ocfmatt** [comment]
Build is successful when fetching Nettle from git with the C99 loop removed.

```console
[root@maxlogin2 spack]# ./bin/spack install nettle
==> libsigsegv is already installed in /opt/software/uoa/2019/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/libsigsegv-2.11-qpmaxx6z62df4s4hwyehdwptv6kmrfhf
==> m4 is already installed in /opt/software/uoa/2019/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/m4-1.4.18-d3ev3mslx5w4xz5tjx4jhhj7sfbv42ub
==> pkgconf is already installed in /opt/software/uoa/2019/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/pkgconf-1.6.0-lb63mhdhdxn77fragqn7f7t2hkunpomb
==> ncurses is already installed in /opt/software/uoa/2019/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/ncurses-6.1-rqw3cmcwvd76h4rjgtaxpvrqg47prwkg
==> readline is already installed in /opt/software/uoa/2019/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/readline-7.0-b4z2xcjalfc4z7tgktx7txhe2my3wlfd
==> gdbm is already installed in /opt/software/uoa/2019/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/gdbm-1.18.1-dv2qksj2dheh7hzmu7iftqhxlkqrinwt
==> perl is already installed in /opt/software/uoa/2019/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/perl-5.26.2-7tmxubxk4qnty7decxqsz5yu5v3p3lqq
==> autoconf is already installed in /opt/software/uoa/2019/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/autoconf-2.69-zzykwjd4qckrmlojghfjk53mayuclyuj
==> automake is already installed in /opt/software/uoa/2019/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/automake-1.16.1-33xp7xdltlskzu4kkd3mqv7gaiqt4hbj
==> libtool is already installed in /opt/software/uoa/2019/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/libtool-2.4.6-bcbiddkmbu4gy7wx6ftntmxqe3c5br3z
==> gmp is already installed in /opt/software/uoa/2019/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/gmp-6.1.2-6bsovvkqwx6zscwtbvjj6egrgizbyycm
==> Installing nettle
==> Searching for binary cache of nettle
==> Warning: No Spack mirrors are currently configured
==> No binary for nettle found: installing from source
==> Cloning git repository: https://github.com/gnutls/nettle.git at commit f5a3a224bf00bef5669366d2ae23c2b2b13b8016
==> No checksum needed when fetching with git
==> Already staged nettle-3.4.1-dzgj2ckkldnwi4e23k6s6zmbanbucbm3 in /opt/software/uoa/2019/spack/var/spack/stage/nettle-3.4.1-dzgj2ckkldnwi4e23k6s6zmbanbucbm3
==> No patches needed for nettle
==> Building nettle [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed nettle
  Fetch: 2.75s.  Build: 17.91s.  Total: 20.65s.
[+] /opt/software/uoa/2019/spack/opt/spack/linux-centos7-x86_64/gcc-4.8.5/nettle-3.4.1-dzgj2ckkldnwi4e23k6s6zmbanbucbm3

```

Other dependencies failed when bumping Nettle to 3.5rc1 making it more favorable to pull the hotfix commit on 3.4.1.

https://github.com/ocfmatt/spack/commit/0c9393eb72f925b978c3eb4e78922691af5ebb04

**Comment by adamjstewart** [comment]
Could we simply download that patch and add it to the Spack package?

**Comment by codeandkey** [comment]
This should have been fixed by #11611 which forces nettle to compile against c99, but it doesn't look like the flag is present in these logs -- is your Spack installation up to date with `develop`?

**Comment by scheibelp** [comment]
@ocfmatt if you are able to reproduce this in spite of https://github.com/spack/spack/issues/11773#issuecomment-503762004 please reopen. Otherwise I consider it resolved.
